### PR TITLE
[Snippets] Deep copy interface

### PR DESCRIPTION
**Pull request by IvanNovoselov**
### Details:
 - *During review of PR18563 it was established that it is hard to perform correct deep copy of LIR, because it owns a range of pointers that in turn needs to be deep copied. In this PR we introduce a convenient and safe way to copy LIR or deep_copy a range of expressions*

### Tickets:
 - *-*
 
### Prerequisites:
- [x] Snippets dynamic pipeline reorganization [18563](https://github.com/openvinotoolkit/openvino/pull/18563)

PR between branches for review: https://github.com/IvanNovoselov/openvino/pull/14

### CODE REVIEW

**Comment by IvanNovoselov** [comment]
@a-sidorova, @v-Golubev could you please start the review?

File: src/common/snippets/src/lowered/loop_manager.cpp [hunk 1]
```diff
@@ -15,8 +15,17 @@
 namespace ov {
 namespace snippets {
 namespace lowered {
+using LoopManager = LinearIR::LoopManager;
+using LoopPort = LoopManager::LoopPort;
+using LoopInfo = LoopManager::LoopInfo;
+
+std::shared_ptr<LoopPort> LoopPort::clone_for_new_expr(const ExpressionPtr& new_expr) const {
```
Comment (a-sidorova) [inline src/common/snippets/src/lowered/loop_manager.cpp:22]: Now we can clone `LoopPort` 🎆  May I ask you to update the WA [there](https://github.com/openvinotoolkit/openvino/blob/3403e6c02873bb0fa9f0cf410d24bbb79991e93c/src/common/snippets/src/lowered/loop_manager.cpp#L323) please?

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 2]
```diff
@@ -49,7 +49,10 @@ class LinearIR {
 
     ExpressionPtr create_expression(const std::shared_ptr<Node>& n, const std::vector<PortConnectorPtr>& inputs);
 
-    static LinearIR::container deep_copy_range(LinearIR::container::const_iterator begin, LinearIR::container::const_iterator end);
+    std::shared_ptr<LinearIR> clone() const;
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp:52]: And for other classes. Because sometimes we can work with only instances. I'd like to avoid situations aka `auto new_smth = *{smth->clone()}` like [there](https://github.com/openvinotoolkit/openvino/blob/f62fd778a5c51ef426138e19f70aa59f1800709d/src/common/snippets/src/op/subgraph.cpp#L494).

File: src/common/snippets/src/lowered/expression.cpp [hunk 3]
```diff
@@ -103,6 +119,54 @@ void Expression::set_loop_ids(const std::vector<size_t>& loops) {
     m_loop_ids = loops;
 }
 
+void Expression::update_node_and_connectors(const ExpressionPtr& expr,
+                                            const std::vector<PortConnectorPtr>& new_inputs,
+                                            const std::shared_ptr<Node>& new_node) {
+    if (new_node) {
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/lowered/expression.cpp:125]: Seems like in `InsertTailLoop`...
If we want to have valid `LinearIR` till code emission, let's make really deep copy for tail loops with node copying. What do you think?

File: src/common/snippets/include/snippets/lowered/expression_port.hpp [hunk 4]
```diff
@@ -25,6 +25,11 @@ class ExpressionPort {
 
     ExpressionPort() = default;
     explicit ExpressionPort(const std::shared_ptr<Expression>& expr, Type type, size_t port);
+    /**
+    * @interface clone
+    * @brief Creates similar Expression port, but for new expression
+     */
+    std::shared_ptr<ExpressionPort> clone_for_new_expr(const std::shared_ptr<Expression>& new_expr) const;
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/expression_port.hpp:32]: Good idea, renamed.

File: src/common/snippets/include/snippets/lowered/expression.hpp [hunk 5]
```diff
@@ -19,12 +19,14 @@ namespace lowered {
...
 class LinearIR;
 using ExpressionPtr = std::shared_ptr<Expression>;
+using ExressionMap = std::unordered_map<Expression*, ExpressionPtr>;
 class Expression : public std::enable_shared_from_this<Expression> {
     friend class LinearIR;
     friend class ExpressionPort;
 
 public:
     Expression() = default;
+    Expression(const Expression& other);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/expression.hpp:29]: We can, but note that it will make `Expressions` non-copyable outside of `Expression::` or friendly classes. This is not a bad thing, since copying generally produces invalid expression (because internal pointers are not updated), but we should keep it in mind.

File: src/common/snippets/src/lowered/expression.cpp [hunk 6]
```diff
@@ -103,6 +119,54 @@ void Expression::set_loop_ids(const std::vector<size_t>& loops) {
     m_loop_ids = loops;
 }
 
+void Expression::update_node_and_connectors(const ExpressionPtr& expr,
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/expression.cpp:122]: Yes, it looks better, let's do it

File: src/common/snippets/src/lowered/expression.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/expression.cpp:125]: The idea here was to allow updates of expression's connections without deep-coping of nGraph IR. For example, if we need to move Expression inside LIR and to change its connectors.
On the other hand, we can always pass `expr->get_node()` to this method, so it's Ok to remove the `if` I guess.

File: src/common/snippets/src/lowered/expression.cpp [hunk 7]
```diff
@@ -103,6 +119,54 @@ void Expression::set_loop_ids(const std::vector<size_t>& loops) {
...
+    OPENVINO_ASSERT(new_inputs.size() == expr->m_input_port_descriptors.size(),
+                    "Can't create Expression with new inputs: invalid number of input port connectors passed");
+    expr->m_input_port_connectors = new_inputs;
+    for (size_t i = 0; i < expr->m_input_port_descriptors.size(); i++) {
+        const auto& i_con = new_inputs[i];
+        const auto& i_port = expr->get_input_port(i);
+        if (!i_con->found_consumer(i_port))
+            i_con->add_consumer(i_port);
+    }
+    auto& new_outputs = expr->m_output_port_connectors;
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/expression.cpp:139]: Theoretically yes, it would be safer, so lets do it
BTW I thought that you're going to suggest `std::transform(...)` 😆

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp:52]: Yes, I see your point, but LIR is a different story. The thing with LIR is that it's handled by value and passed to transformations by reference almost everywhere in the pipeline. So when we write `auto a = m_linear_ir`, we actually expect a shallow copy, not a deep one. For example here, in [op::Kernel](https://github.com/openvinotoolkit/openvino/blob/9fafcabb7c7b221821b0b77616d9967c8d13223f/src/common/snippets/include/snippets/op/kernel.hpp#L23-L26), we don't need to deep copy of the whole IR (which is really expensive). 
In fact, we need to deep copy LIR only once, and that's the place you linked. And it looks a bit unnatural just because `LinearIR::clone()` is aligned with the rest of the `clone` methods and returns `shared_ptr` of the object, while LIR (despite the rest of the classes) is managed by value, not by pointer.

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 8]
```diff
@@ -44,6 +44,35 @@ LinearIR::LinearIR(const std::shared_ptr<ov::Model>& model, const std::shared_pt
...
+    cloned->m_config = m_config;
+    // Clone underlying ngraph nodes
+    NodeVector original_nodes;
+    original_nodes.reserve(m_expressions.size());
+    for (const auto& expr : m_expressions)
+        original_nodes.push_back(expr->get_node());
+    // node_map and expr_map map original node pointer (expression) to a new pointer (expression)
+    ngraph::NodeMap node_map;
+    OPENVINO_SUPPRESS_DEPRECATED_START
+    ngraph::clone_nodes(original_nodes,  node_map);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/linear_ir.cpp:58]: 👍

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 1 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/loop_manager.cpp:22]: I tried, but please double-check. I'm not 100% confident that I got it right 😅

File: src/common/snippets/src/lowered/expression.cpp [hunk 9]
```diff
@@ -103,6 +119,54 @@ void Expression::set_loop_ids(const std::vector<size_t>& loops) {
...
+        expr->m_source_node = new_node;
+    }
+    OPENVINO_ASSERT(new_inputs.size() == expr->m_input_port_descriptors.size(),
+                    "Can't create Expression with new inputs: invalid number of input port connectors passed");
+    expr->m_input_port_connectors = new_inputs;
+    for (size_t i = 0; i < expr->m_input_port_descriptors.size(); i++) {
+        const auto& i_con = new_inputs[i];
+        const auto& i_port = expr->get_input_port(i);
+        if (!i_con->found_consumer(i_port))
+            i_con->add_consumer(i_port);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/expression.cpp]: As discussed offline, it's a good idea, but should be implemented in a separate PR. I've created a dedicated ticket **122690**

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp:52]: I got it, good explained! Thanks a lot!

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 1 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/loop_manager.cpp]: You did exactly what I meant. Thanks!

**Review by a-sidorova — APPROVED** [review]

File: src/common/snippets/src/lowered/expression_factory.cpp [hunk 10]
```diff
@@ -91,7 +91,7 @@ ExpressionPtr LinearIR::ExpressionFactory::create(const std::shared_ptr<op::Loop
                                                   const std::vector<PortConnectorPtr>& inputs,
                                                   const LinearIR& linear_ir) {
     OPENVINO_ASSERT(inputs.empty(), "LoopBegin cannot have inputs");
-    auto expr = std::make_shared<Expression>(Expression(n, linear_ir.m_shape_infer_factory));
+    auto expr = std::shared_ptr<Expression>(new Expression(n, linear_ir.m_shape_infer_factory));
```
Comment (v-Golubev) [inline src/common/snippets/src/lowered/expression_factory.cpp:94]: What is the reason for this change? We can't use `make_shared` for expressions anymore?

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 11]
```diff
@@ -44,6 +44,35 @@ LinearIR::LinearIR(const std::shared_ptr<ov::Model>& model, const std::shared_pt
...
+    // Clone underlying ngraph nodes
+    NodeVector original_nodes;
+    original_nodes.reserve(m_expressions.size());
+    for (const auto& expr : m_expressions)
+        original_nodes.push_back(expr->get_node());
+    // node_map and expr_map map original node pointer (expression) to a new pointer (expression)
+    ngraph::NodeMap node_map;
+    OPENVINO_SUPPRESS_DEPRECATED_START
+    ngraph::clone_nodes(original_nodes, node_map);
+    OPENVINO_SUPPRESS_DEPRECATED_END
```
Comment (v-Golubev) [inline src/common/snippets/src/lowered/linear_ir.cpp:59]: It looks like I missed something: why do we need to clone `ov::Node` here if we clone it inside `deep_copy_range` too?

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 1 above)
Comment (v-Golubev) [reply src/common/snippets/src/lowered/loop_manager.cpp]: @IvanNovoselov  @a-sidorova Actually we can't just use `clone_for_new_expr` instead of old logic because 
new expression ports in `update_loop_port` might have different `port_idx` besides expression. Let's look at a specific example. E.g. we use `update_loops_port` when we insert Load operation on 1st Add's input: in this case new ExpressionPort has idx=0 whereas Add's ExpressionPort has idx=1.

So we can return old logic or introduce a new API for LoopPort: `clone_with_new_expr_port`

File: src/common/snippets/src/lowered/loop_manager.cpp [hunk 12]
```diff
@@ -319,13 +357,11 @@ void LinearIR::LoopManager::update_loop_port(size_t loop_id, const ExpressionPor
...
-                   [](LoopPort loop_port, const ExpressionPort& expr_port) {
-                       LoopPort copy = std::move(loop_port);  // to save loop port parameters
-                       copy.expr_port = std::make_shared<ExpressionPort>(expr_port);
-                       return copy;
-                   });
+    std::vector<LoopPort> target_loop_ports;
+    target_loop_ports.reserve(target_ports.size());
+    for (const auto& p : target_ports)
+        target_loop_ports.push_back(*port_it->clone_with_new_expr(p.get_expr()));
+
```
Comment (v-Golubev) [inline src/common/snippets/src/lowered/loop_manager.cpp:364]: Please note that we need to change not only expression but port_idx too. See more detailed explanations in the [discussion](https://github.com/openvinotoolkit/openvino/pull/20242#discussion_r1354963045)

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 1 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/loop_manager.cpp]: @v-Golubev Oh, I see your point. Great catch! Thanks a lot!

@IvanNovoselov I'm sorry that I confused you. Could you please revert this change? It's invalid since we call `clone_for_new_expr` with new expr that isn't the same as previous expr (they may have different input count, output count). I think that in this case we should add `clone_with_new_expr_port`. You may just revert this change and don't introduce new API with `clone_with_new_expr_port`. It's just small suggestion.

File: src/common/snippets/src/lowered/expression_factory.cpp (hunk 10 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/expression_factory.cpp:94]: The reason is that we implemented a [protected copy constructor](https://github.com/IvanNovoselov/openvino/blob/1e43b55755396b62f3227b9a33860f2a787fdefa/src/common/snippets/include/snippets/lowered/expression.hpp#L64) instead of a default public one.

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 13]
```diff
@@ -44,6 +44,35 @@ LinearIR::LinearIR(const std::shared_ptr<ov::Model>& model, const std::shared_pt
     m_shape_infer = std::make_shared<LIRShapeInfer>(m_expressions, m_io_expressions);
 }
 
+std::shared_ptr<LinearIR> LinearIR::clone() const {
+    auto cloned = std::make_shared<LinearIR>();
+    cloned->m_config = m_config;
+    // Clone underlying ngraph nodes
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/linear_ir.cpp:50]: Even from comments? Wow, that's harsh 🙂 
You can delete it from my comments, but you would never delete it from my heart 🤣

File: src/common/snippets/src/lowered/linear_ir.cpp (hunk 11 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/linear_ir.cpp:59]: No-no, you're right it was there well before a proper deep_copy_range was implemented, but we don't need it anymore. Thanks a lot!

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 1 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/loop_manager.cpp]: All right, I reverted the changes

File: src/common/snippets/src/lowered/loop_manager.cpp (hunk 12 above)
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/loop_manager.cpp:364]: Agreed. Good job!

File: src/common/snippets/src/lowered/expression_factory.cpp (hunk 10 above)
Comment (v-Golubev) [reply src/common/snippets/src/lowered/expression_factory.cpp:94]: Thanks for the explanation

**Comment by v-Golubev** [comment]
BTW do we really have dependency from https://github.com/openvinotoolkit/openvino/pull/18563? When I merged "Deep copy interface" changes with my branch (which do not contain #18563 changes), I did not meet any difficulties. If we did not have the dependency, we would probably merge this PR faster since it's quite small in comparison with #18563

**Comment by IvanNovoselov** [comment]
> BTW do we really have dependency from #18563? When I merged "Deep copy interface" changes with my branch (which do not contain #18563 changes), I did not meet any difficulties. If we did not have the dependency, we would probably merge this PR faster since it's quite small in comparison with #18563

In principle, we can merge it before #18563, since this PR changes are quite isolated. But I would not suggest it because I expect that rebasing #18563 over this PR would be error-prone and could be painful as well.

**Comment by IvanNovoselov** [comment]
@a-sidorova, @v-Golubev thanks a lot for your comments!
@dmitry-gorokhov could you take a look please?

**Review by dmitry-gorokhov — APPROVED** [review]
